fix(core): correctly handle `null` or `undefined` in `ErrorHandler#handleError()`

gkalpak opened the pull request:
Since `ErrorHandler#handleError()` expects an argument of type `any` it should be able to handle values such as `null` and `undefined`. Previously, it failed to handle these values, because it was trying to access properties on them.

This PR fixes it by ensuring no properties are accessed on `null` or `undefined` values.

NOTE: This is part of fully addressing #28106.

Fixes #21252.

JoostK commented:
Nice! This also addresses #21252.

gkalpak commented:
Cool, I didn't know about #21252 😅

mary-poppins commented:
You can preview bc425ef at https://pr42881-bc425ef.ngbuilds.io/.

AndrewKushnir commented on packages/core/src/error_handler.ts:
```diff
@@ -58,17 +59,17 @@ export class ErrorHandler {
-  _findOriginalError(error: Error): any {
-    let e = getOriginalError(error);
+  _findOriginalError(error: any): Error|null {
+    let e = error && getOriginalError(error);
```
I believe that would also be a problem if we pass something other primitive types here like `true` or `10`. This check would succeed when the logic inside of the `getOriginalError` function will fail. Should we have sure that it's a type of object instead?

gkalpak replied:
Reading properties off of other primitives values (other than `null` or `undefined`) is fine (i.e. it doesn't through an error, it just returns `undefined`). I have added more tests to cover that.

AndrewKushnir replied:
Yeah, you are right! Thanks for adding extra tests 👍

AndrewKushnir commented on packages/core/src/errors.ts:
```diff
@@ -21,8 +21,8 @@ export function getOriginalError(error: Error): Error {
-export function getErrorLogger(error: Error): (console: Console, ...values: any[]) => void {
-  return (error as any)[ERROR_LOGGER] || defaultErrorLogger;
+export function getErrorLogger(error: unknown): (console: Console, ...values: any[]) => void {
+  return error && (error as any)[ERROR_LOGGER] || defaultErrorLogger;
```
Similar situation here: we can have `true` or a number here and it'd cause `error[ERROR_LOGGER]` read to fail.

gkalpak replied:
Reading properties off of other primitives values (other than `null` or `undefined`) is fine (i.e. it doesn't through an error, it just returns `undefined`). I have added more tests to cover that.

gkalpak committed:
fix(core): correctly handle `null` or `undefined` in `ErrorHandler#handleError()`

Since `ErrorHandler#handleError()` expects an argument of type `any` it
should be able to handle values such as `null` and `undefined`.
Previously, it failed to handle these values, because it was trying to
access properties on them.

This commit fixes it by ensuring no properties are accessed on `null` or
`undefined` values.

NOTE: This is part of fully addressing #28106.

Fixes #21252

gkalpak committed:
fixup! fix(core): correctly handle `null` or `undefined` in `ErrorHandler#handleError()`

mary-poppins commented:
You can preview 0d96e56 at https://pr42881-0d96e56.ngbuilds.io/.

AndrewKushnir commented:
[Presubmit](https://test.corp.google.com/ui#id=OCL:385793906:BASE:385793928:1626795542062:128df398).

AndrewKushnir commented:
@gkalpak FYI, presubmits went well, adding this PR to the merge queue.
